### ISSUE TITLE
Streaming while caching

### ISSUE THREAD

**Issue by eshkoliGilad** [danikula/AndroidVideoCache#17]
Caching work great and thank you for this lib, but it doesn't seem to stream (playing) while caching.
It looks like it wait for caching to be available at 100% and only then start to play the video. (Which is longer than just stream in via URL and start playing the first chunks available).

Is it normal behavior or is it a malformed code I've wrote ?

```
HttpProxyCacheServer proxy = VideoProxyFactory.getProxy(this);
        proxy.registerCacheListener(new CacheListener() {
            @Override
            public void onCacheAvailable(File cacheFile, String url, int percentsAvailable) {
                if(percentsAvailable<100)
                {
                    mPercentageOfVideo.setText(percentsAvailable+"%");
                    mPercentageOfVideo.setVisibility(View.VISIBLE);
                }
                else
                {
                    mPercentageOfVideo.setText("100%");
                    mVideoSharePath=cacheFile.getAbsolutePath();
                }
            }
        }, mVideoUrl);
        mVideoUrl = proxy.getProxyUrl(mVideoUrl);
```

**Comment by danikula** [comment]
@eshkoliGilad You are wrong. Video starts to play as soon as proxy download anything. See sample of using `sample\Single Video`. In example primary progress shows video's current position and secondary progress shows downloading progress. But if you **seek** video player to position that proxy havn't achieved yet video player will wait. It is no so good but this solution allows to keep proxy implementation very simple.

Please, let me know, if proxy works on your device in some another way as I described above .

**Comment by eshkoliGilad** [comment]
Don't know why you closed this, but I'm using OnPreparedListener with my VideoView and when I use the above code it is getting called only when cache is 100% finished and then and only then it starts to play the video.

**Comment by danikula** [comment]
@eshkoliGilad is this bug reproduced in sample app?

**Comment by baderkhan** [comment]
We are having the same exact issue. We spent the last 4 hours trying to work with the code to make it work. The video doesn't play until the entire video is cached

we noticed that the variable offsetLength (not sure if that is the exact name) from GetRequest is set to the total size of video. This causes ProxyCache read() function to download entire video before it returns. Input stream from the socket contains the "Range" field with value equal to size of video. How can we change that?

**Comment by danikula** [comment]
@baderkhan which video client do you use (VideoView, exoPlayer, smth another)? What the request this client send to proxy? You can see full request in this [line of code](https://github.com/danikula/AndroidVideoCache/blob/master/library/src/main/java/com/danikula/videocache/GetRequest.java#L29).

**Comment by baderkhan** [comment]
We're using the Android MediaPlayer class. This is the initialization for it.

<pre>
Map<String, String> headers = new HashMap();
HttpProxyCacheServer proxy = newProxy();
String proxyUrl = proxy.getProxyUrl(strFileName);
this.m_MediaPlayer.setDataSource(this.m_UnityActivity, Uri.parse(proxyUrl));
</pre>

The MediaPlayer calls PrepareAsync() after that. And we implement OnPreparedListener which gets triggered once the video is ready to play.

The problem is it OnPrepared never gets triggered until the video is 100% downloaded. We have put a debug statement for the constructor of GetRequest. Here are the logs

<pre>
Proxy cache server started. Ping it...
D/ProxyCache( 1369): Open connection  to http://127.0.0.1:53460/ping
D/ProxyCache( 1369): Accept new socket Socket[address=/127.0.0.1,port=50885,localPort=53460]
E/ProxyCache( 1369): GetRequest START:  GET /ping HTTP/1.1
E/ProxyCache( 1369): User-Agent: Dalvik/2.1.0 (Linux; U; Android 5.0.2; SM-G920T1 Build/LRX22G)
E/ProxyCache( 1369): Host: 127.0.0.1:53460
E/ProxyCache( 1369): Connection: Keep-Alive
E/ProxyCache( 1369): Accept-Encoding: gzip
E/ProxyCache( 1369):  GetRequest END
I/ProxyCache( 1369): Request to cache proxy:GetRequest{uri='ping', rangeOffset=0, partial=false}
D/ProxyCache( 1369): Opened connections: 0
D/ProxyCache( 1369): Ping response: `ping ok`, pinged? true
D/ProxyCache( 1369): Accept new socket Socket[address=/127.0.0.1,port=37735,localPort=53460]
E/ProxyCache( 1369): GetRequest START:  GET /http%3A%2F%2Fvideo-us-east-1.youvisit.com%2F241897%2F2048_3d_5.mp4 HTTP/1.1
E/ProxyCache( 1369): User-Agent: Samsung SM-G920T1 stagefright/Beyonce/1.1.9 (Linux;Android 5.0.2)
E/ProxyCache( 1369): Host: 127.0.0.1:53460
E/ProxyCache( 1369): Connection: Keep-Alive
E/ProxyCache( 1369): Accept-Encoding: gzip
E/ProxyCache( 1369):  GetRequest END
I/ProxyCache( 1369): Request to cache proxy:GetRequest{uri='http%3A%2F%2Fvideo-us-east-1.youvisit.com%2F241897%2F2048_3d_5.mp4', rangeOffset=0, partial=false}
D/ProxyCache( 1369): Open connection  to http://video-us-east-1.youvisit.com/241897/2048_3d_5.mp4
D/ProxyCache( 1369): Shutdown proxy for HttpUrlSource{url='http://video-us-east-1.youvisit.com/241897/2048_3d_5.mp4}
D/ProxyCache( 1369): Client communication problem. It seems client closed connection
D/ProxyCache( 1369): Error closing client's input stream: it seems client closed connection
D/ProxyCache( 1369): Opened connections: 0
D/ProxyCache( 1369): Accept new socket Socket[address=/127.0.0.1,port=44392,localPort=53460]
E/ProxyCache( 1369): GetRequest START: GET /http%3A%2F%2Fvideo-us-east-1.youvisit.com%2F241897%2F2048_3d_5.mp4 HTTP/1.1
E/ProxyCache( 1369): User-Agent: Samsung SM-G920T1 stagefright/Beyonce/1.1.9 (Linux;Android 5.0.2)
E/ProxyCache( 1369): Range: bytes=125288404-
E/ProxyCache( 1369): Host: 127.0.0.1:53460
E/ProxyCache( 1369): Connection: Keep-Alive
E/ProxyCache( 1369): Accept-Encoding: gzip
E/ProxyCache( 1369):  GetRequest END
I/ProxyCache( 1369): Request to cache proxy:GetRequest{uri='http%3A%2F%2Fvideo-us-east-1.youvisit.com%2F241897%2F2048_3d_5.mp4', rangeOffset=125288404, partial=true}
D/ProxyCache( 1369): Open connection  with offset 2374952 to http://video-us-east-1.youvisit.com/241897/2048_3d_5.mp4
D/ProxyCache( 1369): Shutdown proxy for HttpUrlSource{url='http://video-us-east-1.youvisit.com/241897/2048_3d_5.mp4}
D/ProxyCache( 1369): Opened connections: 0
D/ProxyCache( 1369): Accept new socket Socket[address=/127.0.0.1,port=47291,localPort=53460]
E/ProxyCache( 1369): GetRequest START: GET /http%3A%2F%2Fvideo-us-east-1.youvisit.com%2F241897%2F2048_3d_5.mp4 HTTP/1.1
E/ProxyCache( 1369): User-Agent: Samsung SM-G920T1 stagefright/Beyonce/1.1.9 (Linux;Android 5.0.2)
E/ProxyCache( 1369): Range: bytes=85877-
E/ProxyCache( 1369): Host: 127.0.0.1:53460
E/ProxyCache( 1369): Connection: Keep-Alive
E/ProxyCache( 1369): Accept-Encoding: gzip
E/ProxyCache( 1369):  GetRequest END:
</pre>

**Comment by danikula** [comment]
@baderkhan Thanks. It is very strange. I don't know why MediaPlayer requests end of file (open request with offset == video's size). 
```
GET /http%3A%2F%2Fvideo-us-east-1.youvisit.com%2F241897%2F2048_3d_5.mp4 HTTP/1.1
User-Agent: Samsung SM-G920T1 stagefright/Beyonce/1.1.9 (Linux;Android 5.0.2)
Range: bytes=125288404-
Host: 127.0.0.1:53460
Connection: Keep-Alive
Accept-Encoding: gzip
```
I'll add workaround for this case.

**Comment by danikula** [comment]
fixed in 2.2.0

**Comment by baderkhan** [comment]
thanks. We will try it and get back
ᐧ

*Taher Baderkhan*
YouVisit LLC
617-821-9722
taher@youvisit.com

[image: Inline image 1] <http://www.youvisit.com/>

**Comment by danikula** [comment]
@baderkhan please check and make feedback. I can't reproduce this issue on my devices, but tests shows that it should work.

**Comment by eshkoliGilad** [comment]
Streaming while caching now works great, however when seeking to a certain point of the video the MediaPlayer throws error exception of: Error playing video:1, -1007. (It is not at all points, but happens always a specific one).

Which means "Bitstream is not conforming to the related coding standard or file spec." according to Android Docs.

This thing does not happen when I'm not using this library. Could it be something related to this issue ?

**Comment by danikula** [comment]
Glad to hear that streaming works. I need some time to investigate problem
you described. Please, create new issue.

вс, 4 окт. 2015, 13:09, Gilad Eshkoli <notifications@github.com>:

> Streaming while caching now works great, however when seeking to a certain
> point of the video the MediaPlayer throws error exception of: Error playing
> video:1, -1007. (It is not at all points, but happens always a specific
> one).
>
> Which means "Bitstream is not conforming to the related coding standard or
> file spec." according to Android Docs.
>
> This thing does not happen when I'm not using this library. Could it be
> something related to this issue ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danikula/AndroidVideoCache/issues/17#issuecomment-145323785>
> .
>
-- 
Best regards,
Alexey Danilov

**Comment by baderkhan** [comment]
The solution you provided in latest release worked perfectly. Thank you
very much for the prompt replay and fix!!!!
ᐧ

*Taher Baderkhan*
YouVisit LLC
617-821-9722
taher@youvisit.com

[image: Inline image 1] <http://www.youvisit.com/>

**Comment by AmanCyber** [comment]
Hi i am using 2.7.0 version, in my case for specific URL its downloading complete video first and then plays, caching works fine. Below is test url not working 
https://storage.googleapis.com/yellfydev/55a3abd76d4b9dc02a128514-1512993998653.mp4?GoogleAccessId=GOOGFJFTVNJQI6ZNTYSJ&Signature=7trIm7OjZe%2BJboAC9eRL7nkHQ7A%3D&Expires=1516602821
 Just mentioning, in this case MD5 name generator not able to find extension, then also cached file is reusable but on first time it downloads complete file first. Also tried with sample provided same is happening.

**Comment by AmanCyber** [comment]
Just for FYI i created my own file name generator still above mentioned problem persists.
